### PR TITLE
[CIR] Add CIRGenerator and plug it via CIRGenAction

### PR DESCRIPTION
**Pull request by lanza**

### CODE REVIEW

File: clang/include/clang/CIRFrontendAction/CIRGenAction.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  enum class OutputType {
+    EmitCIR,
+  };
+
+private:
+  friend class CIRGenConsumer;
+
+  // TODO: this is redundant but just using the OwningModuleRef requires more of
+  // clang against MLIR. Hide this somewhere else.
+  std::unique_ptr<mlir::OwningOpRef<mlir::ModuleOp>> mlirModule;
```
Comment (bcardosolopes) [inline clang/include/clang/CIRFrontendAction/CIRGenAction.h:33]: The original comment is from 2 years ago, any idea why is still relevant?

File: clang/include/clang/CIRFrontendAction/CIRGenAction.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+  mlir::MLIRContext *mlirContext;
+
+protected:
+  CIRGenAction(OutputType action, mlir::MLIRContext *mlirContext = nullptr);
+
+  void foo() {
```
Comment (bcardosolopes) [inline clang/include/clang/CIRFrontendAction/CIRGenAction.h:40]: What is this about?

File: clang/lib/CIR/CodeGen/CIRGenModule.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+#include "clang/Basic/CodeGenOptions.h"
+#include "clang/Basic/Diagnostic.h"
+#include "mlir/IR/MLIRContext.h"
+
+using namespace clang;
+namespace cir {
+/// Implementation of a CIR/MLIR emission from Clang AST.
+///
+/// This will emit operations that are specific to C(++)/ObjC(++) language,
+/// preserving the semantics of the language and (hopefully) allow to perform
```
Comment (bcardosolopes) [inline clang/lib/CIR/CodeGen/CIRGenModule.h:28]: perhaps remove the "hopefully" and just have a more solid statement

File: clang/lib/CIR/CodeGen/CIRGenTypeCache.h [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+#define LLVM_CLANG_LIB_CIR_CODEGENTYPECACHE_H
+
+namespace cir {
+
+/// This structure provides a set of types that are commonly used
+/// during IR emission. It's initialized once in CodeGenModule's
+/// constructor and then copied around into new CIRGenFunction's.
+struct CIRGenTypeCache {
+  CIRGenTypeCache() {}
+
```
Comment (bcardosolopes) [inline clang/lib/CIR/CodeGen/CIRGenTypeCache.h:23]: Tie this up?

**Review by bcardosolopes — COMMENTED** [review]
Perhaps add a `-emit-cir` testcase that doesn't test anything but just proves the wiring works. More comments inline.

File: clang/lib/CIR/CodeGen/CIRGenTypeCache.h (hunk 4 above)
Comment (lanza) [reply clang/lib/CIR/CodeGen/CIRGenTypeCache.h]: what?

File: clang/include/clang/Driver/Options.td [hunk 5]
```diff
@@ -2900,7 +2900,7 @@ defm clangir : BoolFOption<"clangir",
   PosFlag<SetTrue, [], [ClangOption, CC1Option], "Use the ClangIR pipeline to compile">,
   NegFlag<SetFalse, [], [ClangOption, CC1Option], "Use the AST -> LLVM pipeline to compile">,
   BothFlags<[], [ClangOption, CC1Option], "">>;
-def emit_cir : Flag<["-"], "emit-cir">, Visibility<[CC1Option]>,
+def emit_cir : Flag<["-"], "emit-cir">, Visibility<[ClangOption, CC1Option]>,
```
Comment (lanza) [inline clang/include/clang/Driver/Options.td:2903]: I had to readd this. The code in `lib/Driver/Driver.cpp:4810`ish needs to know the file output type from flags to the driver. So we need **a driver flag**. And if we need a driver flag then we might as well merge it with one that correctly feeds the emission type along as well.

File: clang/lib/CIR/CodeGen/CIRGenTypeCache.h (hunk 4 above)
Comment (bcardosolopes) [reply clang/lib/CIR/CodeGen/CIRGenTypeCache.h]: Remove the extra newline

File: clang/test/CIR/hello.c [hunk 6]
```diff
@@ -0,0 +1,5 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir -emit-cir %s | FileCheck %s
+
+// CHECK: CIRGenModule::buildTopLevelDecl
+
+void foo() {}
```
[the remark this answers is not in the document]
Comment (bcardosolopes) [reply clang/test/CIR/hello.c:5]: Oh, I see what you are doing. I was thinking even more simple, no code, just an empty file without invoking FileCheck. So you can remove the printing hack. Before this PR, calling into `-emit-cir` would crash, so the idea of the simple test is to just make sure it doesn't anymore.

File: clang/include/clang/Driver/Options.td (hunk 5 above)
Comment (bcardosolopes) [reply clang/include/clang/Driver/Options.td:2903]: Hmmm, this is content for Aaron, who was opposed to my take.

**Review by bcardosolopes — CHANGES_REQUESTED** [review]

**Pull request closed without merging**